cirq/api/google/v2/device.proto

```diff
@@ -0,0 +1,116 @@
+syntax = "proto3";
+
+package cirq.api.google.v2;
+
+option java_package = "com.google.cirq.api.google.v2";
+option java_outer_classname = "DeviceProto";
+option java_multiple_files = true;
+
+// This contains information about a device that includes the
+// qubits on the device, supported gates, connections, and timing.
+// This message specifies information that is needed when sending a
+// Program message to the device.
+message DeviceSpecification {
+
+  // A list of allowed gatesets for programs submitted to this processor
+  // Language.gate_set should be one of these values to be valid.
+  repeated GateSet valid_gate_sets = 1;
+
+  // A list of allowed ids for qubits within the Program.
+  // Any programs with ids not in this list will be rejected.
+  // If empty, all qubit values are allowed (e.g. in a simulator)
+  repeated string valid_qubits = 2;
+
+  // A list of targets that gates can use.
+  repeated TargetList valid_targets = 3;
+
+  // Additional recommendations, caveats, and soft requirements that
+  // are advice to users of the device, specified in English text
+  // For instance, "All Z gates are converted to VirtualZ gates".
+  string developer_recommendations = 4;
+}
+
+message GateSet {
+   // The name of the gate set corresponding to Language.gate_set
+   string name = 1;
+
+   // A list of valid gates permitted by this gate set
+   repeated GateDefinition valid_gates = 2;
+}
+
+message GateDefinition {
+  // The name for the gate.  This must match the Gate.id in your program.
+  string id = 1;
+
+  // If unset or set to zero, any number of qubits is allowed.
+  int32 number_of_qubits = 2;
+
+  // The name of the arguments that should be specified for
+  // an operation of this gate
+  repeated ArgDefinition valid_args = 3;
+
+  // This defines the approximate amount of time for each gate,
+  // specified as an integer number of picoseconds.
+  int64 gate_duration_picos = 4;
+
+  // Valid targets that this gate can use.
+  // Values in this list correspond to the name of the TargetList
+  // If unset, all combinations with number_of_qubits target are allowed.
+  repeated string valid_targets = 5;
```

Comment:
How would one specify targets for a gate that can take variable numbers of qubits, such as measurement gates, if you want to restrict to a subset of qubits?

Reply:
Made this a SUBSET_PERMUTATION TargetOrdering